File: packages/evals/tasks/bench/observe/observe_taxes.ts
import { defineBenchTask } from "../../../framework/defineTask.js";
import { selectorsResolveToSameElement } from "../../../framework/observeSelectors.js";

export default defineBenchTask(
  { name: "observe_taxes" },
  async ({ debugUrl, sessionUrl, stagehand, page, logger }) => {
    try {
      await page.goto("https://file.1040.com/estimate/");

      const { data: observations } = await stagehand.observe(
        "Find all the form input elements under the 'Income' section",
      );

      if (observations.length === 0) {
        return {
          _success: false,
          observations,
          debugUrl,
          sessionUrl,
          logs: logger.getLogs(),
        };
      } else if (observations.length < 13) {
        return {
          _success: false,
          observations,
          debugUrl,
          sessionUrl,
          logs: logger.getLogs(),
        };
      }

      let foundMatch = false;
      for (const observation of observations) {
        try {
          if (await selectorsResolveToSameElement(page, observation.selector, ["#tpWages"])) {
            foundMatch = true;
            break;
          }
        } catch (error) {
          console.warn(
            `Failed to check observation with selector ${observation.selector}:`,
            String(error),
          );
          continue;
        }
      }

      return {
        _success: foundMatch,
        expected: "#tpWages",
        observations,
        debugUrl,
        sessionUrl,
        logs: logger.getLogs(),
      };
    } catch (error) {
      return {
        _success: false,
        error: String(error),
        debugUrl,
        sessionUrl,
        logs: logger.getLogs(),
      };
    }
  },
);

File: packages/sdk-ruby/evals/tasks/observe/observe_taxes.rb
# frozen_string_literal: true

# Port of packages/evals/tasks/bench/observe/observe_taxes.ts

Evals.define_task("observe_taxes") do |t|
  t.page.goto("https://file.1040.com/estimate/")

  observations = t.stagehand.observe("Find all the form input elements under the 'Income' section").data

  if observations.empty? || observations.length < 13
    next { _success: false, observations: observations.map(&:to_wire) }
  end

  expected_locator = "#tpWages"

  expected_result = t.page.locator(expected_locator).first.inner_text

  found_match = observations.any? do |observation|
    t.page.locator(observation.selector).first.inner_text == expected_result
  rescue StandardError => e
    t.logger.log("Failed to check observation with selector #{observation.selector}: #{e.message}")
    false
  end

  { _success: found_match, expected: expected_result, observations: observations.map(&:to_wire) }
end
